trinity/algorithm/advantage_fn/grpo_advantage.py

```diff
@@ -81,3 +87,121 @@ def default_args(cls) -> Dict:
+                (https://arxiv.org/pdf/2506.02355).
+        """
+        self.epsilon = epsilon
+        self.std_threshold = std_threshold
+        self.duplicate_experiences = duplicate_experiences
+        self.rank_penalty = rank_penalty
+
+    def group_experiences(self, exps):
+        return group_by(exps, id_type="task")
+
+    def calculate_group_advantage(
+        self, group_id: str, exps: List[Experience]
+    ) -> Tuple[List[Experience], Dict]:
+        metrics = {}
+        with torch.no_grad():
+            if len(exps) == 1:
+                group_reward_mean = torch.tensor(0.0)
+                group_reward_std = torch.tensor(1.0)  # set to 1.0 to avoid division by zero
+                if self.std_threshold is not None:
+                    metrics["skipped_count"] = 1
+                    exps.clear()  # Clear experiences if only one experience
+            else:
+                rewards = torch.tensor([exp.reward for exp in exps], dtype=torch.float32)
+
+                if self.rank_penalty is not None:
+                    # Correct bias by adjusting rewards based on their ranks
+                    old_log_probs = torch.tensor(
+                        [torch.mean(exp.logprobs, axis=-1) for exp in exps]
+                    )
+                    group_ranks = torch.argsort(torch.argsort(old_log_probs))
+                    group_ranks = group_ranks / len(group_ranks)
+                    rewards = rewards * (1 - group_ranks * self.rank_penalty)
+
+                group_reward_mean = torch.mean(rewards)
+                group_reward_std = torch.std(rewards)
+
+                # If the reward standard deviation is below a threshold, skip the group
+                if self.std_threshold is not None and group_reward_std <= self.std_threshold:
+                    metrics["skipped_count"] = len(exps)
+                    exps.clear()
+
+            for exp in exps:
+                score = (exp.reward - group_reward_mean) / (group_reward_std + self.epsilon)
+                exp.advantages = score * exp.action_mask
+                exp.returns = exp.advantages.clone()
+
+            metrics["reward_mean"] = group_reward_mean.item()
+            metrics["reward_std"] = group_reward_std.item()
+
+        return exps, metrics
+
+    def _duplicate_experiences(self, exp_groups: Dict[str, List[Experience]]) -> List[Experience]:
+        original_group_num = len(exp_groups)
+        exp_groups = {
+            group_id: group_exps
+            for group_id, group_exps in exp_groups.items()
+            if len(group_exps) > 0
+        }
+        skipped_group_num = original_group_num - len(exp_groups)
+        if skipped_group_num == 0:
+            return [exp for group in exp_groups.values() for exp in group]
+        elif skipped_group_num == original_group_num:
+            # All groups are skipped, return an empty list
+            return []
+        else:
+            idx = torch.randperm(len(exp_groups))[:skipped_group_num]
```

Comment:
```suggestion
            idx = torch.randperm(len(exp_groups))[:skipped_group_num].tolist()
```
It seems tensor cannot serve as index